File: backend/claimMapper.js
const currencyFields = new Set([
  'Charge_Amount',
  'Allowed_Amount',
  'Paid_Amount',
  'Patient_Responsibility',
  'Adjustment_Amount',
])

function parseAmount(value) {
  const numberValue = Number(value)
  return Number.isFinite(numberValue) ? Number(numberValue.toFixed(2)) : 0
}

function parseInteger(value) {
  const numberValue = Number.parseInt(value, 10)
  return Number.isFinite(numberValue) ? numberValue : 0
}

function parseEdiDate(value) {
  const text = String(value || '').trim()
  if (!/^\d{8}$/.test(text)) return ''
  return `${text.slice(0, 4)}-${text.slice(4, 6)}-${text.slice(6, 8)}`
}

function parseEdiTimestamp(value) {
  const text = String(value || '').trim()
  if (!/^\d{14}$/.test(text)) return ''

  const year = text.slice(0, 4)
  const month = text.slice(4, 6)
  const day = text.slice(6, 8)
  let hour = Number(text.slice(8, 10))
  const minute = text.slice(10, 12)
  const period = hour >= 12 ? 'PM' : 'AM'
  hour = hour % 12 || 12

  return `${year}-${month}-${day} ${hour}:${minute} ${period}`
}

function formatClaimNumber(claimId) {
  const digits = String(claimId || '').replace(/\D/g, '')
  return `CLM-${digits.slice(-6).padStart(6, '0')}`
}

function cleanRawRow(row) {
  return Object.fromEntries(
    Object.entries(row).map(([key, value]) => {
      if (currencyFields.has(key)) return [key, parseAmount(value)]
      if (key === 'Units') return [key, parseInteger(value)]
      return [key, typeof value === 'string' ? value.trim() : value]
    }),
  )
}

export function normalizeClaim(row) {
  const firstName = row.Patient_First_Name?.trim() || ''
  const lastName = row.Patient_Last_Name?.trim() || ''
  const claimId = row.Claim_ID?.trim() || ''

  return {
    claimId,
    number: formatClaimNumber(claimId),
    memberClaimNumber: row.Claim_Number_For_Member?.trim() || '',
    memberId: row.Member_ID?.trim() || '',
    groupId: row.Group_ID?.trim() || '',
    groupName: row.Group_Name?.trim() || '',
    payer: row.Payer_Name?.trim() || '',
    payerId: row.Payer_ID?.trim() || '',
    patient: [firstName, lastName].filter(Boolean).join(' '),
    patientFirstName: firstName,
    patientLastName: lastName,
    dob: parseEdiDate(row.Patient_DOB),
    gender: row.Patient_Gender?.trim() || '',
    accountNumber: row.Patient_Account_Number?.trim() || '',
    subscriberId: row.Subscriber_Member_ID?.trim() || '',
    billingProviderNpi: row.Billing_Provider_NPI?.trim() || '',
    billingProvider: row.Billing_Provider_Name?.trim() || '',
    renderingProviderNpi: row.Rendering_Provider_NPI?.trim() || '',
    renderingProvider: row.Rendering_Provider_Name?.trim() || '',
    dos: parseEdiDate(row.Service_Date_From),
    serviceEnd: parseEdiDate(row.Service_Date_To),
    placeOfServiceCode: row.Place_of_Service_Code?.trim() || '',
    placeOfService: row.Place_of_Service_Description?.trim() || '',
    cptCode: row.CPT_Code?.trim() || '',
    cptDescription: row.CPT_Description?.trim() || '',
    diagnosisCode: row.ICD10_Diagnosis_Code?.trim() || '',
    diagnosisDescription: row.ICD10_Diagnosis_Description?.trim() || '',
    units: parseInteger(row.Units),
    totalCharge: parseAmount(row.Charge_Amount),
    allowed: parseAmount(row.Allowed_Amount),
    paid: parseAmount(row.Paid_Amount),
    patientResp: parseAmount(row.Patient_Responsibility),
    adjustment: parseAmount(row.Adjustment_Amount),
    statusCode: row.Claim_Status_Code?.trim() || '',
    status: row.Claim_Status_Description?.trim() || '',
    denialReason: row.Denial_Reason?.trim() || '',
    filingIndicator: row.Claim_Filing_Indicator?.trim() || '',
    priorAuth: row.Prior_Auth_Number?.trim() || '',
    referral: row.Referral_Number?.trim() || '',
    transactionVersion: row.HIPAA_Transaction_Version?.trim() || '',
    submissionDate: parseEdiDate(row.Submission_Date),
    createdAt: parseEdiTimestamp(row.Created_Timestamp),
    edi: {
      isaControlNumber: row.ISA_Control_Number?.trim() || '',
      gsControlNumber: row.GS_Control_Number?.trim() || '',
      stTransactionSetId: row.ST_Transaction_Set_ID?.trim() || '',
    },
    raw: cleanRawRow(row),
    importedAt: new Date(),
  }
}

export function buildMemberDocuments(claims) {
  const grouped = new Map()

  for (const claim of claims) {
    const existing = grouped.get(claim.memberId) || []
    existing.push(claim)
    grouped.set(claim.memberId, existing)
  }

  return Array.from(grouped.entries()).map(([memberId, memberClaims]) => {
    const sortedClaims = [...memberClaims].sort((a, b) => b.dos.localeCompare(a.dos))
    const latestClaim = sortedClaims[0]
    const totals = sortedClaims.reduce((acc, claim) => {
      acc.totalCharges += claim.totalCharge
      acc.totalAllowed += claim.allowed
      acc.totalPaid += claim.paid
      acc.totalPatientResp += claim.patientResp
      acc.totalAdjustment += claim.adjustment
      if (claim.status === 'Denied') acc.deniedClaimCount += 1
      return acc
    }, {
      totalCharges: 0,
      totalAllowed: 0,
      totalPaid: 0,
      totalPatientResp: 0,
      totalAdjustment: 0,
      deniedClaimCount: 0,
    })

    return {
      memberId,
      patient: latestClaim.patient,
      patientFirstName: latestClaim.patientFirstName,
      patientLastName: latestClaim.patientLastName,
      dob: latestClaim.dob,
      gender: latestClaim.gender,
      subscriberId: latestClaim.subscriberId,
      groupId: latestClaim.groupId,
      groupName: latestClaim.groupName,
      primaryPayer: latestClaim.payer,
      latestServiceDate: latestClaim.dos,
      latestClaimNumber: latestClaim.number,
      billingProviders: [...new Set(sortedClaims.map((claim) => claim.billingProvider).filter(Boolean))],
      payers: [...new Set(sortedClaims.map((claim) => claim.payer).filter(Boolean))],
      claimCount: sortedClaims.length,
      totalCharges: Number(totals.totalCharges.toFixed(2)),
      totalAllowed: Number(totals.totalAllowed.toFixed(2)),
      totalPaid: Number(totals.totalPaid.toFixed(2)),
      totalPatientResp: Number(totals.totalPatientResp.toFixed(2)),
      totalAdjustment: Number(totals.totalAdjustment.toFixed(2)),
      deniedClaimCount: totals.deniedClaimCount,
      importedAt: new Date(),
    }
  })
}

File: backend/claim_mapper.py
from datetime import datetime, timezone

CURRENCY_FIELDS = {
    "Charge_Amount",
    "Allowed_Amount",
    "Paid_Amount",
    "Patient_Responsibility",
    "Adjustment_Amount",
}


def parse_amount(value):
    try:
        return round(float(value or 0), 2)
    except (TypeError, ValueError):
        return 0


def parse_integer(value):
    try:
        return int(value or 0)
    except (TypeError, ValueError):
        return 0


def parse_edi_date(value):
    text = str(value or "").strip()
    if len(text) != 8 or not text.isdigit():
        return ""
    return f"{text[:4]}-{text[4:6]}-{text[6:8]}"


def parse_edi_timestamp(value):
    text = str(value or "").strip()
    if len(text) != 14 or not text.isdigit():
        return ""

    hour = int(text[8:10])
    minute = text[10:12]
    period = "PM" if hour >= 12 else "AM"
    hour = hour % 12 or 12
    return f"{text[:4]}-{text[4:6]}-{text[6:8]} {hour}:{minute} {period}"


def format_claim_number(claim_id):
    digits = "".join(character for character in str(claim_id or "") if character.isdigit())
    return f"CLM-{digits[-6:].rjust(6, '0')}"


def clean_raw_row(row):
    cleaned = {}
    for key, value in row.items():
        if key in CURRENCY_FIELDS:
            cleaned[key] = parse_amount(value)
        elif key == "Units":
            cleaned[key] = parse_integer(value)
        else:
            cleaned[key] = value.strip() if isinstance(value, str) else value
    return cleaned


def normalize_claim(row):
    first_name = str(row.get("Patient_First_Name", "") or "").strip()
    last_name = str(row.get("Patient_Last_Name", "") or "").strip()
    claim_id = str(row.get("Claim_ID", "") or "").strip()

    return {
        "claimId": claim_id,
        "number": format_claim_number(claim_id),
        "memberClaimNumber": str(row.get("Claim_Number_For_Member", "") or "").strip(),
        "memberId": str(row.get("Member_ID", "") or "").strip(),
        "groupId": str(row.get("Group_ID", "") or "").strip(),
        "groupName": str(row.get("Group_Name", "") or "").strip(),
        "payer": str(row.get("Payer_Name", "") or "").strip(),
        "payerId": str(row.get("Payer_ID", "") or "").strip(),
        "patient": " ".join(part for part in [first_name, last_name] if part),
        "patientFirstName": first_name,
        "patientLastName": last_name,
        "dob": parse_edi_date(row.get("Patient_DOB")),
        "gender": str(row.get("Patient_Gender", "") or "").strip(),
        "accountNumber": str(row.get("Patient_Account_Number", "") or "").strip(),
        "subscriberId": str(row.get("Subscriber_Member_ID", "") or "").strip(),
        "billingProviderNpi": str(row.get("Billing_Provider_NPI", "") or "").strip(),
        "billingProvider": str(row.get("Billing_Provider_Name", "") or "").strip(),
        "renderingProviderNpi": str(row.get("Rendering_Provider_NPI", "") or "").strip(),
        "renderingProvider": str(row.get("Rendering_Provider_Name", "") or "").strip(),
        "dos": parse_edi_date(row.get("Service_Date_From")),
        "serviceEnd": parse_edi_date(row.get("Service_Date_To")),
        "placeOfServiceCode": str(row.get("Place_of_Service_Code", "") or "").strip(),
        "placeOfService": str(row.get("Place_of_Service_Description", "") or "").strip(),
        "cptCode": str(row.get("CPT_Code", "") or "").strip(),
        "cptDescription": str(row.get("CPT_Description", "") or "").strip(),
        "diagnosisCode": str(row.get("ICD10_Diagnosis_Code", "") or "").strip(),
        "diagnosisDescription": str(row.get("ICD10_Diagnosis_Description", "") or "").strip(),
        "units": parse_integer(row.get("Units")),
        "totalCharge": parse_amount(row.get("Charge_Amount")),
        "allowed": parse_amount(row.get("Allowed_Amount")),
        "paid": parse_amount(row.get("Paid_Amount")),
        "patientResp": parse_amount(row.get("Patient_Responsibility")),
        "adjustment": parse_amount(row.get("Adjustment_Amount")),
        "statusCode": str(row.get("Claim_Status_Code", "") or "").strip(),
        "status": str(row.get("Claim_Status_Description", "") or "").strip(),
        "denialReason": str(row.get("Denial_Reason", "") or "").strip(),
        "filingIndicator": str(row.get("Claim_Filing_Indicator", "") or "").strip(),
        "priorAuth": str(row.get("Prior_Auth_Number", "") or "").strip(),
        "referral": str(row.get("Referral_Number", "") or "").strip(),
        "transactionVersion": str(row.get("HIPAA_Transaction_Version", "") or "").strip(),
        "submissionDate": parse_edi_date(row.get("Submission_Date")),
        "createdAt": parse_edi_timestamp(row.get("Created_Timestamp")),
        "edi": {
            "isaControlNumber": str(row.get("ISA_Control_Number", "") or "").strip(),
            "gsControlNumber": str(row.get("GS_Control_Number", "") or "").strip(),
            "stTransactionSetId": str(row.get("ST_Transaction_Set_ID", "") or "").strip(),
        },
        "raw": clean_raw_row(row),
        "importedAt": datetime.now(timezone.utc),
    }


def build_member_documents(claims):
    grouped = {}
    for claim in claims:
        grouped.setdefault(claim["memberId"], []).append(claim)

    documents = []
    for member_id, member_claims in grouped.items():
        sorted_claims = sorted(member_claims, key=lambda claim: claim.get("dos", ""), reverse=True)
        latest_claim = sorted_claims[0]
        denied_count = sum(1 for claim in sorted_claims if claim.get("status") == "Denied")

        documents.append({
            "memberId": member_id,
            "patient": latest_claim.get("patient", ""),
            "patientFirstName": latest_claim.get("patientFirstName", ""),
            "patientLastName": latest_claim.get("patientLastName", ""),
            "dob": latest_claim.get("dob", ""),
            "gender": latest_claim.get("gender", ""),
            "subscriberId": latest_claim.get("subscriberId", ""),
            "groupId": latest_claim.get("groupId", ""),
            "groupName": latest_claim.get("groupName", ""),
            "primaryPayer": latest_claim.get("payer", ""),
            "latestServiceDate": latest_claim.get("dos", ""),
            "latestClaimNumber": latest_claim.get("number", ""),
            "billingProviders": sorted({claim.get("billingProvider", "") for claim in sorted_claims if claim.get("billingProvider")}),
            "payers": sorted({claim.get("payer", "") for claim in sorted_claims if claim.get("payer")}),
            "claimCount": len(sorted_claims),
            "totalCharges": round(sum(claim.get("totalCharge", 0) for claim in sorted_claims), 2),
            "totalAllowed": round(sum(claim.get("allowed", 0) for claim in sorted_claims), 2),
            "totalPaid": round(sum(claim.get("paid", 0) for claim in sorted_claims), 2),
            "totalPatientResp": round(sum(claim.get("patientResp", 0) for claim in sorted_claims), 2),
            "totalAdjustment": round(sum(claim.get("adjustment", 0) for claim in sorted_claims), 2),
            "deniedClaimCount": denied_count,
            "importedAt": datetime.now(timezone.utc),
        })

    return documents
